File: content/courses/parallel_computing_introduction/solns/bcast_ex.f90
program broadcaster
    use mpi
    implicit none

    integer           :: nprocs, rank, ierr
    integer, dimension(:), allocatable :: values
    integer           :: i

    call MPI_Init(ierr)
    call MPI_Comm_size(MPI_COMM_WORLD,nprocs,ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD,rank,ierr)

    allocate(values(10))

    if ( rank==0 ) then
        values=[(i,i=1,10)]
    endif

    call MPI_Bcast(values,size(values),MPI_INTEGER,0,MPI_COMM_WORLD,ierr)

    write(*,'(11i5)'), rank, values

    call MPI_Finalize(ierr)

end program 

File: content/courses/parallel-computing-introduction/solns/bcast_ex.cxx
#include <iostream>
#include <iomanip>
#include "mpi.h"

using namespace std;

int main(int argc, char *argv[]) {

    int nprocs,rank;

    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    int *values=new int[10];
    for (int i=1; i<=10; ++i){
        values[i-1]=i;
    }

    MPI_Bcast(values,10,MPI_INT,0,MPI_COMM_WORLD);

    cout<<rank<<" "<<values[0];
    for (int i=1; i<9; ++i) {
       cout<<" "<<values[i];
    }
    cout<<" "<<values[9]<<endl;

    MPI_Finalize();

}
        

